kernel/src/main/java/org/kframework/kompile/DefinitionParsing.java

```diff
@@ -269,69 +271,76 @@ protected Definition resolveConfigBubbles(Definition definition, Module defaultC
-                            return Configuration(items.get(0), BooleanUtils.TRUE, configContents.att());
-                        case "#ruleEnsures":
-                            return Configuration(items.get(0), items.get(1), configContents.att());
-                        default:
-                            throw KEMException.compilerError("Illegal configuration with requires clause detected.", configContents);
-                        }
-                    })
-                    .flatMap(
-                            configDecl -> stream(GenerateSentencesFromConfigDecl.gen(configDecl.body(), configDecl.ensures(), configDecl.att(), parser.getExtensionModule(), kore)))
-                    .collect(Collections.toSet());
-        }
+                Set<Sentence> parsedSet = stream(m.localSentences())
+                        .parallel()
+                        .filter(s -> s instanceof Bubble && ((Bubble) s).sentenceType().equals(configuration))
+                        .map(b -> (Bubble) b)
+                        .flatMap(b -> performParse(cache.getCache(), parser, parser.getScanner(options.global), b)
+                                .map(p -> upSentence(p, b.sentenceType())))
+                        .collect(Collections.toSet());
 
-        Module mapModule;
-        if (def.getModule("MAP").isDefined()) {
-            mapModule = def.getModule("MAP").get();
-        } else {
-            throw KEMException.compilerError("Module Map must be visible at the configuration declaration, in module " + module.name());
-        }
-        return Module(module.name(), (Set<Module>) module.imports().$bar(Set(mapModule)),
-                (Set<Sentence>) module.localSentences().$bar(configDeclProductions)
-                        .filter(s -> !(s instanceof Bubble && ((Bubble) s).sentenceType().equals(configuration))),
-                module.att());
+                return Module(m.name(), m.imports(),
+                        stream((Set<Sentence>) m.localSentences().$bar(parsedSet)).filter(s -> !(s instanceof Bubble && ((Bubble) s).sentenceType().equals(configuration))).collect(Collections.toSet()), m.att());
+            }
+        }, "parse config bubbles").apply(defWithCaches);
+
+        return DefinitionTransformer.from(m -> {
+            if (stream(m.localSentences()).noneMatch(s -> s instanceof Configuration))
+              return m;
+
+            Set<Sentence> importedConfigurationSortsSubsortedToCell = stream(m.productions())
+                  .filter(p -> p.att().contains("cell"))
+                  .map(p -> Production(Seq(), Sorts.Cell(), Seq(NonTerminal(p.sort())))).collect(toSet());
+
+            Module module = Module(m.name(), m.imports(),
+                  (Set<Sentence>) m.localSentences().$bar(importedConfigurationSortsSubsortedToCell),
+                  m.att());
+
+            Set<Sentence> configDeclProductions;
+            ParseCache cache = loadCache(gen.getConfigGrammar(module));
```

Comment:
This and the following line are expensive steps, and you're doing them twice. It would be better if we could avoid redoing this work.